tags/release-0.9.1/pMars/doc/pmars.6
.\" portable MARS is
.\" Copyright (c) by Albert Ma, Na'ndor Sieben, Stefan Strack, Mintardjo Wangsaw and Ilmari Karonen
.\"
.\" pmars.6 by Stefan Strack
.\" $Id: pmars.6,v 1.1.1.1 2005/09/11 21:55:12 alexanderwait Exp $
.\"
.\" to create a printable manual with proper pagination do
.\" nroff -man -e pmars.6 > pmars.doc
.\" ex pmars.doc
.\" :%s/_\^H//g
.\" :g/PMARS(6)/-2,-1d
.\" :g/PMARS(6)/+1d
.\" :g/Last change:/-1d
.\" :g/Last change:/+1,+2d
.\" :wq
.de EX      \"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +5n
..
.de EE      \"End example
.fi
.in -5n
.if n .sp 1
.if t .sp .5
..
.TH PMARS 6 "December 25, 2000" "PMARS v0.9.2"
.ad b
.SH NAME
pmars \- portable corewar system with ICWS'94 extensions
.SH SYNOPSIS
.I pmars
[
.I option \fP...\fI
]
file1
[
.I file(s)
]
.SH DESCRIPTION
.I pMARS
(portable Memory Array Redcode Simulator) is a corewar
interpreter with multi-platform support.
.I pMARS
currently runs on UNIX systems, PC/DOS, VMS, Amiga (AmigaDOS command line),
and the Mac.
.I pMARS
implements the ICWS'94 draft standard, but can also be used in
ICWS'88 mode. The base system includes a graphical core
display for UNIX (curses, X-windows), PC/linux (svgalib), PC/DOS and the Mac
(see APPENDIX). A line-oriented debugger is
included to help in writing warriors for the ICWS'94 draft standard.
.PP
.I pMARS
runs one or more warriors written in
.I Redcode
that are provided as
.I file(s)
on the command line.
Running a single warrior is supported for debugging. Two warriors are
pitted against each other for standard play, but up to 36 warriors can be
named for "multi-warrior" core war. If the warrior(s)
assemble without error they are loaded into the core array and
executed in round-robin mode starting with the first warrior.
Warrior 1 is loaded starting at core position
0, warrior 2, 3, etc., at either a random or fixed position.
For fairness the starting order is rotated after each round.
.SH SCORE
The score is reported after all rounds
have been played. A round ends when either
a single surviving warrior remains or when a maximum number of cycles
has elapsed.  For each round, every surviving warrior is awarded
points calculated from a score formula (F).
By default, this is (W*W-1)/S, where
W is the total number of warriors participating, S is the number of
of survivors, and "/" is integer division. Alternative score formulas can be
specified with the
.I   =
option (see below).
.PP
When two warriors battle, the results are reported as wins1, wins2, and ties,
where wins1 and wins2 are the numbers of rounds that each warrior
won, and ties are the number of rounds in which both warriors lasted until time-out.
The score is then:
.PP
.nf
        warrior 1:  points = wins1 * F + ties * F
        warrior 2:  points = wins2 * F + ties * F
.fi
.PP
F is a score formula containing the W and S parameters.
When more than two warriors (W) participate in a game, results are reported as
wins[1], wins[2], .., wins[W], losses for each warrior, where the index
indicates the number of warriors that survived until the end (S parameter in the score formula). The total number of points for each warrior is then calculated as:
.PP
.nf
        points =  sum (S=1..W) (wins[S] * F)
.fi
.PP
A few alternative score formulas:
.EX
        10
        (W+1)/S
        (x=W-S+1)*(x+1)/2
        (S==3)*5 + (S==2)*3 + (S==1)
        (S == 1)*W + (S != 1)
.EE
.SH OPTIONS
Command line options may occur anywhere on the command line. Single-letter
options without
a parameter can be combined as in \-fbe. The special argument
.I -
(dash) stands for standard input (stdin). It can be combined with the
.I -@
option (see below) to signify reading options from stdin, or the
.I -
can take the place of a  warrior filename, in which case warrior code
starting with a
.I ;redcode
line and ending with an
.I END
statement is extracted from stdin. The
.I END
statement can be omitted if the next
.I ;redcode
line immediately follows the last instruction. Several warriors, each specified
by a separate dash on the command line and bracketed by
.I ;redcode/END
can be piped into
.I pMARS.
.I "#-"
(where # is a positive number) is a shorthand for writing this number of
dashes on the command line.
.TP
.BI \-r\ #
This options sets the number of rounds to play. The default is 1. \-r 0
produces assembly output (unless -b is specified), but does not execute
the warrior(s). A maximum of 32787 rounds can be specified.
.TP
.BI \-s\ #
The \-s option specifies the size of core memory in number of instructions.
It defaults to 8000. Maximum core size is platform-dependent, but usually
at least 65535.
.TP
.BI \-c\ #
\-c sets the maximum number of cycles per round.
A cycle consists of one execution of each warrior. A round ends when either a
single warrior remains executing or when the maximum number of cycles has
elapsed.
.TP
.BI \-p\ #
This option sets the maximum number of processes a warrior can run.
The default is 8000.
.TP
.BI \-l\ #
This sets the maximum length of a warrior source file in instructions.
It defaults to 100 and can be up to 500.
.TP
.BI \-d\ #
This option specifies the minimum distance between the first instruction of
each warrior. It cannot be smaller than the maximum length (-l option) and
defaults to 100 instructions.
.TP
.BI \-S\ #
The -S option sets the size of the P-space to # cells. The default is 1/16th
of core size if core size is evenly divisible by sixteen, or the next larger
divisible fraction. See also the P-SPACE section below.
.TP
.BI \-f
The \-f option instructs the loader to use a fixed series of addresses for
positioning warriors. This is done by initializing the pseudo random number
generator with a checksum value derived from the source of all warriors.
Thus, initial placements will still be "random" from round to round,
but will be the same if the same warriors are run again. As a
consequence, the result of battles
run with the \-f option will show no statistical
fluctuations. This options is useful for validating ports of
.I pMARS
to new platforms and for providing an absolute, albeit arbitrary
performance measure for warriors.
.TP
.BI \-F\ #
This option server two purposes. It sets the seed value of the pseudo
random number generator, and also tells the loader to install warrior
2 at the given address # for round 1. If the number is larger than the
core size minus the minimum warrior distance, it will be wrapped
around to the range of available starting positions.
This option is useful for testing different
versions of a warrior with constant initial placement.
Warrior 1 is always installed at address 0.
The \-F and \-f options are mutually
exclusive. If neither option is specified, the pseudo random number
generator is initialized with the system time.
.TP
.BI \-P
This option makes
.I pMARS
use an alternative algorithm for positioning warriors. This algorithm uses
all possible combinations of starting order and position in a random
sequence, and will not use the same combination twice unless all the
combinations have already been used. If the \-r option is not given, the
number of rounds fought will be the number of all possible combinations.
Trying to use this option with more (or less) than two warriors causes an
error.
.TP
.BI \-e
If this option is specified, the
.I cdb
debugger is entered immediately after
loading warriors into core.
.TP
.BI \-b
This options runs
.I pMARS
in brief output mode. \-b suppresses warrior listings after assembly.
.TP
.BI \-k
With the \-k option,
.I pMARS
uses the same output format as the KotH program. This option enables
.I pMARS
to work with scripts written for the KotH server.
.TP
.BI \-8
This options enforces strict compliance with the ICWS'88 standard and
disables all ICWS'94 language extensions, which are flagged as syntax errors
by the assembler. Since ICWS'94 is a superset of ICWS'88, this options is
not necessary to run ICWS'88 warriors.
.TP
.BI \-o
When this option is given,
.I pMARS
reports scores in decreasing order rather
than in the order warriors were named on the command line. This is mostly
useful for multi-warrior play.
.TP
.BI \-V
The assembler generates verbose output with this option. This is only useful
for debugging.
.TP
.BI \-v
This option sets the display mode for UNIX and DOS display versions (see
APPENDIX).
.TP
.BI \-@\ <fn>
.I pMARS
continues reading options and filenames from the parameter file
.I <fn>.
If
.I <fn>
is a - (dash) standard input is parsed. Comments in the parameter file start
with a semicolon.
.TP
.BI -=\ <string>
The = (equal) option allows you to specify a custom score formula (see
SCORE). The formula may contain the standard arithmetic and logical operators
(as in the C language), as well as the parameters W (number of warriors
participating) and S (number of warriors surviving this round). You need to
enclose the formula string with quotes if it contains spaces or characters
interpreted by the operating system.
.TP
.BI \-Q\ #
The "query" option is intended for use in scripts that run
.I pMARS
in batch mode. Depending on the number code following
.I -Q,
the program returns an informative number as the exit status.
"pmars -Q 1000" e.g. will cause
.I pMARS
to exit with a code that spells out the program version. More on
.I -Q
arguments and exit codes can be found in the ADDENDUM.
.TP
.BI $
The $ (dollar) parameter is not preceded by a dash and cannot be grouped with
other options. It terminates reading command line parameters and is used in
parameter files or input streams
.I (-@ fn).
The
.I $
is necessary if you want to combine two or more of command line parameters,
warrior code and cdb commands in the same input stream or file. Below an example
of such a combined input file:
.EX
    ;Below the command line parameters:
    -r 10 -beF 1000 2- $
    ;redcode
    ;name Imp 1
    ;assert 1
    mov 0,1
    end
    ;redcode
    ;name Imp 2
    ;assert 1
    mov 0,2
    mov 0,2
    end
    !! cdb commands follow:
    sk 1000
    reg
    quit
.EE
.SH X-WINDOWS OPTIONS
The X-Windows display version of pMARS has these additional command line
options:
.PP
.TP
.BI \-display\ <string>
Expects a string specifying the display where the window should appear.
Following X standards, this display specification has a format of
hostname.domain:display.screen where the part before the colon is the
standard internet host specification and display and screen are integers.
The screen specification can be omitted. For an example, consider you
are working at an X terminal named ncd13.complang.tuwien.ac.at and
remotely logged in at host stud1.tuwien.ac.at where the binary of
pmars lies, you can use the following command line
.PP
.nf
  stud1$ pmars -display ncd13.complang.tuwien.ac.at:0 -b aeka.red aeka.red
.fi
  and the window will appear at you local screen.
.TP
.BI \-geometry\ <string>
Lets you specify the initial size and position of the window. The format
is widthxheight+x+y where width, height, x, y are integers and + may be
replaced by -. Either of the two parts (widthxheight or +x+y) may be
omitted. This overrides the -v switch concerning the window
geometry. As an example, pmars -geometry 600x400+30+100 ....
will open a window of 600 by 400 pixels at the screen position (30,100).
pmars -geometry 1024x768 ... creates a window of 1024x768 pixels and
pmars -geometry -20+300 ... creates a window with standard size with its
left upper corner at position (-20,300), i.e. out of the left side of the
screen.
.TP
.BI \-fn\ <string>
The string following this argument specifies the X font to use. By
default, a terminal font is used. If this font can't be found, "fixed"
is used for a font ("fixed" should be present at every X installation).
Use the command 'xlsfonts' to get a listing of fonts that can be used
at your X server. 'xfontsel' provides a comfortable way to select a
font. "fixed" is the fallback if the specified font can't be found.
It is strongly recommended to use a fixed pitch font.
.PP
The argument of the
.I -v
display option has an additional fourth digit in the X-Windows version: 0 
(the default)
enables the color display, 1 the grayscale, and 2 the black and white display
(e.g. -v 1033 for the grayscale display).
.SH REDCODE
.I pMARS
implements an extension of the proposed ICWS'94 standard. The new instruction
modifiers
.I .A,.B,.AB,.F,.X,
and
.I .I,
the arithmetic instructions
.I MUL, DIV
and
.I MOD,
as well as post-increment indirect (>) are supported.
.I pMARS
currently does not implement read/write ranges, although they may be
added in future versions.
.LP
Version 0.5 of
.I pMARS
adds support for three experimental opcodes that are currently not included
in the ICWS'94 draft:
.LP
SEQ (Skip if EQual): this is a synonym for CMP and is included mainly for
clarity (future versions of
.I pMARS
may implement SEQ as a "predefined EQU" rather than a true opcode).
.LP
SNE (Skip if Not Equal): the opposite of SEQ.
.LP
NOP (No OPerations): do nothing.
.LP
Version 0.6 adds three new indirect addressing modes that use the A-field
instead of the B-field as the pointer for indirection:
.LP
* - indirect using A-field
.LP
{ - predrecement indirect using A-field
.LP
} - postincrement indirect using A-field
.PP
The new P-space instructions of version 0.8 are discussed under separate
heading below.
.PP
The assembler also supports multi\-line
.I EQU
statements, a feature not included in the current ICWS'94 draft.
The format for multi-line EQUates is
.EX
    <label> EQU <line1>
            EQU <line2>
            [...]
            EQU <lineN>
.EE
<label> in the warrior source is replaced by <line1> <newline> <line2>
<newline> [....] lineN. In contrast to KotH, pmars
.I EQUs
substitute arbitrary text, and not just expressions.
.I  EQU
expressions are not implicitly parenthesized.
.LP
.I pMARS
features the non-standard
.I FOR
text-repetition macro. 
This macro repeats the text between
.I FOR
and
.I ROF
a specified number of times:
.EX
<labels> <counter> FOR <times>
                   [..]
                   ROF
.EE
<times> is an expression specifying the number of expansions; it may
contain EQUates and labels as long as they have been defined before the
FOR/ROF block.
<counter> is the last label before the FOR word, but not necessarily on the
same line.
It is expanded to 01, 02, ..,
<times> in each repetition. The optional <labels> label the first
instruction after
.I FOR
expansion. An example:
.EX
        ORG start
start
sp      FOR 2
a&sp        SPL a&sp
            JMP a&sp
        ROF

becomes after expansion

start
a01     SPL a01
        JMP a01
a02     SPL a02
        JMP a02
.EE
The symbol & concatenates 'a' and 01, 02 to form a valid label.
.I EQU
expansion and FOR/ROF processing is done in the same pass.
It is therefore possible to write
.EX
dest01  EQU 500
dest02  EQU 1000
dest03  EQU 1500

idx     FOR 3
            MOV src,dest&idx
            MOV src+1,dest&idx
        ROF

src     MOV <-1,<-1
        JMP src,<-2
.EE
Using predefined EQUates (see below) it is possible to define adaptive FOR/ROF
blocks. The next example fills the remainder of the warrior (up to MAXLENGTH lines)
with decoy instructions:
.EX
    FOR MAXLENGTH-CURLINE
        DAT 1,1
    ROF
.EE
Since true logical expressions
have a value of "1" and false expressions a value of "0", you can write conditionally 
assembled code enclosed by FOR/ROF:
.EX
    FOR CORESIZE == 8000
    <some code>
    ROF

    FOR CORESIZE != 8000
    <other code>
    ROF
.EE
.I pMARS
uses KotH-style
.I ;name
and
.I ;author
comments to describe warriors.
If a line starting with ;redcode is present, all text preceding it is
ignored. This makes it possible to run posted warriors without removing
mail headers and descriptive text.
The
.I ;version, ;strategy
and other comments mentioned in the ICWS'94 draft are currently not used.
.LP
As another "non-standard" extension, the assembler predefines the following
run-time variables: CORESIZE, MAXPROCESSES, MAXCYCLES, MAXLENGTH,
MINDISTANCE, ROUNDS, and PSPACESIZE.
They can be used in your
.I Redcode
as though they were defined by EQUs like
.EX
    CORESIZE     EQU 55440  ;current value of -s parameter
    MAXPROCESSES EQU 10000  ;current value of -p parameter
    [etc.]
.EE
The run-time variable CURLINE holds the current instruction number offset
from the first instruction. WARRIORS is initialized with the number of warriors
specified on the command line.
.LP
.I pMARS supports the
.I ;assert
directive as a way of checking whether a warrior is run under the parameters
it was designed for. If the expression following
.I ;assert
evaluates to "0" (false), assembly is aborted with an error message. If an
.I ;assert
is missing, a warning is issued. Examples:
.EX
    ;assert CORESIZE == 55440 && MAXLENGTH >= 200
    ;assert !(CORESIZE % 4)  ; is multiple of 4
    ;assert 1 ; if warrior works under all settings
.EE
The run-time variable VERSION holds the current
.I pMARS
version (e.g. "60" is v0.6.0) and is useful in ;assert expressions.
.\".PP
.\"RANDOM is initialized to a pseudo-random value between 0 and CORESIZE-1
.\"at the start of each assembly, i.e. multiple RANDOM's in your warrior
.\"source reference the same number.  RANDOM can be used randomize boot
.\"distances to thwart specialized attacks, but it is generally not a good
.\"idea to overuse RANDOM as this makes scores less predictable and improving
.\"a warrior design more difficult.
.\"A possible use for RANDOM is for empirically determining optimal scanning
.\"or bombing constants.
.PP
With the \-8 option,
.I pMARS
is fully ICWS'88 compatible, except that a comma is required between
operands. Extensions like predefined and multi-line EQUs and FOR/ROF are
supported even in ICWS'88 mode.
.LP
A full treatment of corewar and ICWS'94 in particular is beyond
the scope of this document. General information about corewar as well as
the ICWS'94 draft is available by anonymous FTP from soda.berkeley.edu in
pub/corewar.
.SH P-SPACE
Originating from discussions on rec.games.corewar, P-space is an
attempt at making warriors more "intelligent" by giving them a memory.
P-space, short for "private", "permanent" or "priviledged" space is a
memory area separate from core whose contents is not cleared between
rounds. Every warrior has its own P-space for gathering information
about the opposing warrior, but there is a provision for sharing
P-space in team play (see below).
.PP
P-space cells contain values in the range 0..CORESIZE-1. The number of
P-space cells can be adjusted with the -S command line option; by
default, P-space size is 1/16th of CORESIZE. This number is available
to warriors as the predefined variable PSPACESIZE.  pMARS updates
P-space cell 0 at the beginning of each round with the result of the
previous round: 0 signifies a loss in the previous round, a number
larger than zero means that the warrior survived until the end of the
round, the value indicating the number of surviving warriors. That is,
a value of "1" means that the warrior survived by itself (a "win" in
a two-warrior battle), a value of "2" that two warriors lasted until
the end (a "tie" in a two warrior battle), etc..
In the first round, P-cell 0 is set to -1 (actually CORESIZE-1) to
indicate that there is no previous result.
.PP

Changes.txt
06/01/2007 - nMars0.8.2 - Refactoring
  - IProject is vehicle for IParser as well for IEngine 
  - Project now contains WarriorFiles and can be saved/loaded to .nmproj , also from pmars config files
  - nMars.exe is new commandline replacement, some new options, hooks for plugins added, configurable components
  - Parser and Engine are no more executables
  - signed assemblies
  - Async Engine moved to RedCode assembly
  - IDE is now showing tootip info about each cell, including last events
  - console debugger is broken now
  - nMars.Engine bugs in instructions fixed
  - Unit test is producing .nmproj setup for failed tests and can be drive by commandline
  - executed 100k matches on combinations of 5k different warriors to test Engine compatibility with pMars. No open issue.
  - started work on documentation comments in code
  - Parser is now returning result structure with list messages, including positions in files
  - pMars.DllWrapper parser is converting error to result structure too, + memory leaks fixed, WARRIORS count is set from rules
  - pMars.dll a-z registry cleanup in dll mode, long filenames buffer overflow fixed
  - class ComponentSetup renamed to ComponentLoader and simplified

12/23/2006 - nMars0.8.1 - IDE
  - IDE framework
  - fast GUI debug watches
  - async engine layer
  - project, solution
  - IDE editor
  - engine layers and iterfaces extended

11/19/2006 - nMars0.8 - debugger
  - console debugger
  - IronPython extension for debugger
  - fixed bugs in engine
  - building step back infrastructure
  - building coredump infrastructure
  
11/13/2006 - nMars0.7 - going public
  - documentation
  - exception handling
  - XML dump
  
11/13/2006 - nMars0.6 - engine unit test
  - refactored engine
  - pMars.dll silent
  - enhanced unit test
    - command line
    - comparer engine
  - dump match results
  - droped module dependencies
  - fixed bugs in engine
     - PSpace support
     - adresssing bugs in engine

11/08/2006 - nMars0.5 - more binaries
  - simplified modules
  - modules loading by name
  - engine commandline
  - IProject introduced
  - version info in binaries
  - release tools in svn
  
11/03/2006 - nMars0.4 - pMars included
  - included pMars circa 0.9.2, with SDL display from M Joonas Pihlaja
  - modularized pMars and wrapped into Windows DLL
  - used pMars.dll as validator for unit tests
  - enhanced SimpleEngine compatiblity, fixed bugs
  - extended interfaces in namespace RedCode

09/28/2006 - nMars0.3 - first binary
  - Parser more pmarsv compatible
  - mono supported
  - comments parsed
  - unit tests
  - pmarsv wrapper
  
09/17/2006 - nMars0.2 - Parser more powerfull
  - more expression operations
  - label can be A B I ... <Modifier>
  - dynamic label names name & order
  - locations to pareser errors
  - 2 pass label&variable evaluation
  - unit test skeleton
  - exaple warriors in directory rc

08/12/2006 - nMars0.1 - First version

trunk/PMars/Src/pmars.c
/* pMARS -- a portable Memory Array Redcode Simulator
 * Copyright (C) 1993-1996 Albert Ma, Na'ndor Sieben, Stefan Strack and Mintardjo Wangsawidjaja
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

/*
 * pmars.c: main(), toplevel, initialization, cleanup
 * $Id: pmars.c,v 1.1.1.1 2000/08/20 13:29:40 iltzu Exp $
 */

#include <stdio.h>
#if defined(unix) || defined(VMS) || defined(SDLGRAPHX)
#include <signal.h>
#else
#if defined(__MSDOS__)
#include <dos.h>
#endif
#endif
#include "global.h"

#if defined(LINUXGRAPHX)
#include <vga.h>
#include <vgamouse.h>
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>
#endif

#if defined(SDLGRAPHX)
/* Including SDL.h here makes sure main() is renamed to whatever SDL needs. */
#include "SDL.h"
#endif

#if defined(DOSTXTGRAPHX) || defined(DOSGRXGRAPHX) || defined(LINUXGRAPHX) \
    || defined(XWINGRAPHX) || defined(SDLGRAPHX)
#if defined(DJGPP)
#include <std.h>
#else
#ifdef DOS16
#include <io.h>
#else
#if defined(HAVE_UNISTD_H)
#include <unistd.h>
#endif
#endif
#endif
#endif

#ifdef __MAC__
#include <console.h>
#endif

#ifdef __OS2__
#include <signal.h>
void _cdecl sighandler(int);
#endif

#ifdef NEW_STYLE
void    init(void);
extern void results(FILE * outp);
void    body(void);
void    Exit(int);
int     returninfo(void);
#ifdef PSPACE
void    pspace_init(void);
void    pspace_finalize(void);
#endif
#if defined(unix) || defined(__MSDOS__) || defined(SDLGRAPHX) || defined(VMS)
void    sighandler(int dummy);
#endif
#if defined(CURSESGRAPHX)
extern void end_curses(void);
#endif
#else
void    init();
extern void results();
void    body();
void    Exit();
int     returninfo();
#ifdef PSPACE
void    pspace_init();
#endif
#if defined(unix) || defined(__MSDOS__) || defined(SDLGRAPHX)
void    sighandler();
#endif
#if defined(CURSESGRAPHX)
extern void end_curses();
#endif
#endif

/* external strings */
extern char *stub386, *info01, *outOfMemory;
#if defined(LINUXGRAPHX)
extern char *cantInitSvga, *cantOpenConsole, *tcgetattrFails;
#endif

/* global variables */
int     pspP;                /* contains number of P-spaces allocated, <
                                 * warriors with PIN */

#if defined(__BORLANDC__) && defined(CHECK386)
/*
 * Check for 386 or better processor; needed when compiled using 386
 * instruction set under BC. Define CHECK386 to enable processor check.
 */
void    check386(void);
#pragma startup check386        /* must be called before main() */
void
check386(void)
{
  asm     pushf
          asm pop ax
          asm and ah, 01111111 b
          asm or ah, 01000000 b
          asm push ax
          asm popf
          asm pushf
          asm pop ax
          asm and ah, 11000000 b
          asm cmp ah, 01000000 b
          asm je is386
          printf(stub386);
  Exit(NOT386);
is386:
  return;
}
#endif

#if defined(DOSTXTGRAPHX) || defined(DOSGRXGRAPHX) || defined(LINUXGRAPHX) \
    || defined(XWINGRAPHX) || defined(SDLGRAPHX) || defined(STDGRAPHX)
void
decode_vopt(option)
  int     option;
{
#if defined(XWINGRAPHX)
  extern int xDisplayType;        /* defined in xwindisp.c */

  xDisplayType = (option / 1000) % 10;
  option %= 1000;
#endif
  displayLevel = option % 10;
  displayMode = (option % 100 - displayLevel) / 10;
  displaySpeed = (option - displayMode * 10 - displayLevel) / 100;
  if (displaySpeed >= SPEEDLEVELS)
    displaySpeed = SPEEDLEVELS - 1;
#if defined(CURSESGRAPHX)
  refreshInterval = refIvalAr[displaySpeed];
#else
  loopDelay = loopDelayAr[displaySpeed];
  keyDelay = keyDelayAr[displaySpeed];
#endif
#if defined(SDLGRAPHX)
  sdlgr_set_displayLevel(displayLevel);
  sdlgr_set_displaySpeed(displaySpeed);
  sdlgr_set_displayMode(displayMode);
#endif
#if defined(STDGRAPHX)
  stdio_set_displaySpeed(displaySpeed);
  stdio_set_displayLevel(displayLevel);
  stdio_set_displayMode(displayMode);
#endif
}
#endif

void
init()
{
  INITIALINST.opcode = (FIELD_T) DAT *8 + (FIELD_T) mF;
  INITIALINST.A_mode = INITIALINST.B_mode = (FIELD_T) DIRECT;
  INITIALINST.A_value = INITIALINST.B_value = INITIALINST.debuginfo = 0;
  errorcode = SUCCESS;
  errorlevel = WARNING;                /* reserve for future */
  errmsg[0] = '\0';                /* reserve for future */
#if defined(DOSTXTGRAPHX) || defined(DOSGRXGRAPHX) || defined(LINUXGRAPHX) \
    || defined(XWINGRAPHX) || defined(SDLGRAPHX) || defined(STDGRAPHX)
  decode_vopt(SWITCH_v);
#endif
#if defined(DOSTXTGRAPHX) || defined(DOSGRXGRAPHX) || defined(LINUXGRAPHX) \
    || defined(XWINGRAPHX) || (defined(SDLGRAPHX) && !defined(WIN32))
  if (!isatty(fileno(stdin)))
    inputRedirection = TRUE;
#endif
}

void body_load()
{
  int     i, j;
  for (i = 0; (i < warriors) && (errorcode == SUCCESS); i++)
    if ((!assemble(warrior[i].fileName, i)) && (!SWITCH_b)) {
      fprintf(STDOUT, info01, warrior[i].name, warrior[i].instLen,
              warrior[i].authorName);
      disasm(warrior[i].instBank, warrior[i].instLen, warrior[i].offset);
      fprintf(STDOUT, "\n");
    }
#ifdef PSPACE                        /* set up pSpace */
  pspace_init();
#endif
}

void body_results()
{
	int     i, j;
    if (SWITCH_k) {
      set_reg('W', (long) warriors);        /* 'W' used in score calculation */
      if (warriors == 2)        /* standard 2-warrior game */
        fprintf(STDOUT, "%d %d\n%d %d\n", warrior[0].score[0],
             warrior[0].score[1], warrior[1].score[0], warrior[1].score[1]);
      else                        /* multiwarrior */
        for (i = 0; i < warriors; i++) {
          fprintf(STDOUT, "%d ", score(i));
          for (j = 0; j < warriors; j++) {
            fprintf(STDOUT, "%d ", warrior[i].score[j]);
          }
          fprintf(STDOUT, "%d\n", deaths(i));
        }
    } else                        /* ! SWITCH_k */
      results(stdout);
}

void body_finalize()
{
	int w ;
	warriors=0;
	for (w = 0; (w < MAXWARRIOR); w++)
	{
		FREE(warrior[w].authorName);
		FREE(warrior[w].fileName);
		FREE(warrior[w].date);
		FREE(warrior[w].version);
		FREE(warrior[w].name);
		FREE(warrior[w].instBank);
		warrior[w].offset=0;
		warrior[w].instLen=0;
		warrior[w].tasks=0;
		warrior[w].pSpaceIDNumber=0;
		warrior[w].lastResult=0;
		warrior[w].taskHead=NULL;
		warrior[w].taskTail=NULL;
		warrior[w].nextWarrior=NULL;
		warrior[w].authorName=NULL;
		warrior[w].fileName=NULL;
		warrior[w].date=NULL;
		warrior[w].version=NULL;
		warrior[w].name=NULL;
		warrior[w].instBank=NULL;
	}
#ifdef PSPACE                        /* teardown up pSpace */
	pspace_finalize();
#endif
}

void
body()
{
  
  body_load();
  if (rounds && (errorcode == SUCCESS)) {
    simulator1();
	body_results();
  }
  body_finalize();
}

/* called when ctrl-c is pressed; prepares for debugger entry */
#if defined(unix) || defined(__MSDOS__) || defined (__OS2__) || defined(SDLGRAPHX)
void
#ifdef __OS2__
        _cdecl
#endif
sighandler(dummy)
  int     dummy;
{
#ifndef SERVER
  debugState = STEP;
  cmdMod = RESET;                /* terminate any macros in process */
  /*
   * some machines (SPARCstation, RS6000) seem to lose the handler address
   * after first signal trap -> need to call signal() again
   */
#if !defined(__OS2__) && !defined(DJGPP)
#ifdef SIGINT
  signal(SIGINT, sighandler);
#endif
#if 0
  signal(SIGTSTP, sighandler);
#endif
#endif                                /* os2 or djgpp */
  return;
#else
  Exit(USERABORT);
#endif
}
#endif                                /* unix */


void
Exit(errorcode)
  int     errorcode;
{
#if defined(CURSESGRAPHX)
  end_curses();                        /* Restore terminal to sane mode */
#else
#if defined(__MAC__)
  while (!button());
#endif
#endif
#ifdef __OS2__                        /* jk - exit() overreacts... could be Borland
                                 * bug */
#ifdef OS2PMGRAPHX
  _endthread();                        /* this should be called ONLY from a
                                 * secondary thread */
#else
  _exit(errorcode);                /* jk - exit() overreacts... could be Borland
                                 * bug */
#endif
#else
#ifdef LINUXGRAPHX
  if (vga_getcurrentmode() != TEXT)
    svga_display_close(0);
  fflush(stdout);
#endif
  exit(SWITCH_Q >= 0 ? returninfo() : errorcode);
#endif
}

int
main(argc, argv)
  int     argc;
  char  **argv;
{
#if defined(unix) && !defined(DJGPP) 
#ifdef SIGINT
  signal(SIGINT, sighandler);
#endif
#if 0
  signal(SIGTSTP, sighandler);
#endif
#else
#if defined(__MSDOS__) && !defined(DJGPP)
  ctrlbrk((int (*) (void)) sighandler);
#endif
#endif                                /* ifdef unix */
#ifdef __OS2__
  signal(SIGINT, sighandler);
#endif
#ifdef __MAC__
  /* console init for MAC */
  extern int ccommand(char ***);

  argc = ccommand(&argv);
#endif
#if defined(LINUXGRAPHX)
  if (vga_init() == -1) {
    fprintf(stderr, cantInitSvga);
    exit(1);
  }
  if ((console_fd = open("/dev/console", O_RDONLY)) == -1) {
    perror(cantOpenConsole);
    exit(1);
  }
  if (tcgetattr(console_fd, &tio_orig) == -1) {
    perror(tcgetattrFails);
    exit(1);
  }
#endif
#if defined(XWINGRAPHX)
  xWinArgc = argc;
  xWinArgv = argv;
#endif

  if ((errorcode = parse_param(argc, argv)) == 0) {
    init();
#ifdef OS2PMGRAPHX                /* jk */
    pm_body();
#else
    body();
#endif
    Exit(errorcode);
  }
  return SWITCH_Q >= 0 ? returninfo() : errorcode;
}

/* return exitcode based on SWITCH_Q setting, useful mainly in scripts */
int
returninfo()
{
#ifdef GRAPHX
#define VARIANT 2
#else
#ifdef SERVER
#define VARIANT 0
#else
#define VARIANT 1
#endif
#endif

  if (SWITCH_Q >= 1 && SWITCH_Q <= 999) {
    int     idxV[MAXWARRIOR], scrV[MAXWARRIOR], i;

    set_reg('W', (long) warriors);        /* 'W' used in score calculation */
    for (i = 0; i < warriors; ++i) {
      scrV[i] = score(i);
      idxV[i] = i;
    }
    if (SWITCH_o)
      sort_by_score(idxV, scrV);
    if (SWITCH_Q <= warriors)
      return scrV[idxV[SWITCH_Q - 1]];
    else if (SWITCH_Q >= 101 && SWITCH_Q <= 100 + warriors)
      return idxV[SWITCH_Q - 101] + 1;
    else
      return -1;                /* unknown -Q argument */
  } else
    switch (SWITCH_Q) {
    case 1000:
      return PMARSVER;                /* version */
    case 1001:
      return VARIANT;                /* SERVER, GRAPHX, etc. */
    case 1002:
      return PMARSVER * 10 + VARIANT;        /* a combination of the two */
    case 1003:                        /* core address size (int or short) */
      return sizeof(ADDR_T);
    case 2000:                        /* how many warriors share a P-space */
#ifdef PSPACE
      return warriors - pspP + 1;
#else
      return -1;
#endif
    default:
      return -1;                /* unknown -Q argument */
    }
}

#ifdef PSPACE
void
pspace_init()
{
  int     i, j;
  pspP=0;
  for (i = 0; i < warriors; ++i) {
    if (warrior[i].pSpaceIndex == UNSHARED)
      warrior[i].pSpaceIndex = pspP++;
    else if (warrior[i].pSpaceIndex == PIN_APPEARED) {        /* PIN appeared and
                                                         * unmatched */
      warrior[i].pSpaceIndex = pspP;
      for (j = i + 1; j < warriors; ++j)
        if (warrior[i].pSpaceIDNumber == warrior[j].pSpaceIDNumber)
          warrior[j].pSpaceIndex = pspP;
      ++pspP;
    }                                /* else: pSpace already shared, don't
                                 * increment pspP */
  }
  /* allocate space */
  for (i = 0; i < warriors; ++i) {
    warrior[i].lastResult = coreSize - 1;        /* unlikely number */
    if (pSpace[j = warrior[i].pSpaceIndex] == NULL &&
#ifdef DOS16
        ((pSpace[j] = (ADDR_T far *) farcalloc(pSpaceSize, sizeof(ADDR_T))) == NULL)) {
#else
    ((pSpace[j] = (ADDR_T *) calloc(pSpaceSize, sizeof(ADDR_T))) == NULL)) {
#endif
      errout(outOfMemory);
      Exit(MEMERR);
    }
  }
}                                /* pspace_init() */

void pspace_finalize()
{
  int     i, j;
  for (i = 0; i < warriors; ++i) {
	j = warrior[i].pSpaceIndex;
    if (pSpace[j])
	{
		free(pSpace[j]);
		pSpace[j]=NULL;
	}
  }
}
#endif
